### ISSUE TITLE
account for implicit "i" factor in special_kz calculations

### ISSUE THREAD

**Issue by stevengj** [NanoComp/meep#1026]
As documented here:

https://github.com/NanoComp/meep/blob/4e1802b653b1949a9cf33d82ce67583091e17e11/src/step_db.cpp#L121-L133

we implicitly store `i*(TM fields)` when doing special_kz calculations with real field variables.  This doesn't affect things like flux, but it should be accounted for in other calculations, e.g. when retrieving/outputting the fields or their Fourier transforms, handling source terms, etcetera.

If we want to make special_kz the default, we should probably force complex fields by default in that case, since otherwise it will be a bit confusing that the fields are real.

**Comment by stevengj** [comment]
Would be good to check that using `special_kz=True` and `force_complex=True` still gives a significant benefit over the current method of using a 3d calculation, hopefully losing only a factor of 2 compared to the 10x speedup in #1010.

**Comment by oskooi** [comment]
Shown below are the timing results for the test in #1010:[comment](https://github.com/NanoComp/meep/pull/1010#issuecomment-530183763) for three test cases: 
1. `special_kz=True` and `force_complex_fields=False` (2d)
2. `special_kz=True` and `force_complex_fields=True` (2d)
3. `special_kz=False` (3d)

**1. `special_kz=True` and `force_complex_fields=False`**
```
Using MPI version 3.1, 1 processes
-----------
Initializing structure...
Halving computational cell along direction y
time for choose_chunkdivision = 0.000217714 s
Working in 2D dimensions.
Computational cell is 14 x 14 x 0 with resolution 50
     block, center = (0,0,0)
          size (1e+20,1,1e+20)
          axes (1,0,0), (0,1,0), (0,0,1)
          dielectric constant epsilon diagonal = (12,12,12)
time for set_epsilon = 0.235485 s
-----------
MPB solved for omega_1(1.03923,0,0.2) = 0.325185 after 16 iters
MPB solved for omega_1(0.947353,0,0.2) = 0.300075 after 11 iters
MPB solved for omega_1(0.947078,0,0.2) = 0.3 after 5 iters
MPB solved for omega_1(0.947078,0,0.2) = 0.3 after 1 iters
Meep progress: 10.25/216.6666717529297 = 4.7% done in 4.0s, 80.6s to go
on time step 1025 (time=10.25), 0.00390602 s/step
Meep progress: 20.73/216.6666717529297 = 9.6% done in 8.0s, 75.7s to go
on time step 2073 (time=20.73), 0.00382017 s/step
Meep progress: 31.26/216.6666717529297 = 14.4% done in 12.0s, 71.2s to go
on time step 3126 (time=31.26), 0.00380133 s/step
Meep progress: 41.75/216.6666717529297 = 19.3% done in 16.0s, 67.1s to go
on time step 4175 (time=41.75), 0.00381598 s/step
.........
.........
.........
Elapsed run time = 84.2137 s
```

**2. `special_kz=True` and `force_complex_fields=True`**
```
Using MPI version 3.1, 1 processes
-----------
Initializing structure...
Halving computational cell along direction y
time for choose_chunkdivision = 0.000206253 s
Working in 2D dimensions.
Computational cell is 14 x 14 x 0 with resolution 50
     block, center = (0,0,0)
          size (1e+20,1,1e+20)
          axes (1,0,0), (0,1,0), (0,0,1)
          dielectric constant epsilon diagonal = (12,12,12)
time for set_epsilon = 0.235298 s
-----------
Meep: using complex fields.
MPB solved for omega_1(1.03923,0,0.2) = 0.325185 after 16 iters
MPB solved for omega_1(0.947353,0,0.2) = 0.300075 after 11 iters
MPB solved for omega_1(0.947078,0,0.2) = 0.3 after 5 iters
MPB solved for omega_1(0.947078,0,0.2) = 0.3 after 1 iters
Meep progress: 4.76/216.6666717529297 = 2.2% done in 4.0s, 178.2s to go
on time step 476 (time=4.76), 0.00840725 s/step
Meep progress: 9.52/216.6666717529297 = 4.4% done in 8.0s, 174.1s to go
on time step 952 (time=9.52), 0.0084038 s/step
Meep progress: 14.35/216.6666717529297 = 6.6% done in 12.0s, 169.3s to go
on time step 1435 (time=14.35), 0.00829405 s/step
Meep progress: 19.18/216.6666717529297 = 8.9% done in 16.0s, 164.8s to go
on time step 1918 (time=19.18), 0.00828451 s/step
Meep progress: 24.0/216.6666717529297 = 11.1% done in 20.0s, 160.7s to go
on time step 2400 (time=24), 0.00831181 s/step
.........
.........
.........
Elapsed run time = 181.1233 s
```

**3. `special_kz=False`**
```
Using MPI version 3.1, 1 processes
-----------
Initializing structure...
Halving computational cell along direction y
time for choose_chunkdivision = 0.000226817 s
Working in 3D dimensions.
Computational cell is 14 x 14 x 0.02 with resolution 50
     block, center = (0,0,0)
          size (1e+20,1,1e+20)
          axes (1,0,0), (0,1,0), (0,0,1)
          dielectric constant epsilon diagonal = (12,12,12)
time for set_epsilon = 0.906314 s
-----------
Meep: using complex fields.
MPB solved for omega_1(1.03923,0,0.2) = 0.325185 after 16 iters
MPB solved for omega_1(0.947353,0,0.2) = 0.300075 after 11 iters
MPB solved for omega_1(0.947078,0,0.2) = 0.3 after 5 iters
MPB solved for omega_1(0.947078,0,0.2) = 0.3 after 1 iters
Meep progress: 0.28/216.6666717529297 = 0.1% done in 4.0s, 3113.3s to go
on time step 28 (time=0.28), 0.143878 s/step
Meep progress: 1.59/216.6666717529297 = 0.7% done in 8.0s, 1088.4s to go
on time step 159 (time=1.59), 0.0306682 s/step
Meep progress: 2.88/216.6666717529297 = 1.3% done in 12.1s, 896.1s to go
on time step 288 (time=2.88), 0.0312039 s/step
Meep progress: 4.15/216.6666717529297 = 1.9% done in 16.1s, 823.3s to go
on time step 415 (time=4.15), 0.0315457 s/step
Meep progress: 5.43/216.6666717529297 = 2.5% done in 20.1s, 782.1s to go
on time step 543 (time=5.43), 0.0314627 s/step
.........
.........
.........
Elapsed run time = 691.1063 s
```

Even with `force_complex_fields=True`, the `special_kz=True` feature enables a speed up of 3.8X relative to `special_kz=False` even though it is 2.2X slower than `force_complex_fields=False` and `special_kz=True`. This would indicate that it is probably worthwhile to turn this feature on by default.
